### PR TITLE
Add Oak Containers integration test

### LINKED ISSUE #4166: Oak Container: Refactor example to use networking instead of VSOCK
This in conjunction with #4149 should allow us to retire gRPC over VSOCK in favor of the more familiar and stable gRPC over network.

### PR DESCRIPTION
**Pull request by conradgrobler**
Fixes #4166

### CODE REVIEW

File: oak_containers_hello_world_untrusted_app/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use oak_containers_launcher::Launcher;
+
+mod app_client;
+
+/// The local address that will be forwarded by the VMM to the guest on port 8080.
+const UNTRUSTED_APP_ADDRESS: &str = "http://127.0.0.1:8088";
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/src/lib.rs:21]: Ideally the local host port would be specified by the caller (or auto assigned -- though auto assignment is problematic because the caller needs to know it, so let's leave that aside for now).
Can we pass it via the launcher args?

File: oak_containers_hello_world_untrusted_app/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    app_client: app_client::TrustedApplicationClient,
+}
+
+impl UntrustedApp {
+    pub async fn create(
+        launcher_args: oak_containers_launcher::Args,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        let launcher = oak_containers_launcher::Launcher::create(launcher_args).await?;
+        // TODO(#4194): Stop sleeping once we have a mechanism to notify the untrusted app that the
+        // tursted app is ready.
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/src/lib.rs:34]:
```suggestion
        // trusted app is ready.
```

File: oak_containers_hello_world_untrusted_app/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    launcher: Launcher,
+    app_client: app_client::TrustedApplicationClient,
+}
+
+impl UntrustedApp {
+    pub async fn create(
+        launcher_args: oak_containers_launcher::Args,
+    ) -> Result<Self, Box<dyn std::error::Error>> {
+        let launcher = oak_containers_launcher::Launcher::create(launcher_args).await?;
+        // TODO(#4194): Stop sleeping once we have a mechanism to notify the untrusted app that the
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/src/lib.rs:33]: A slightly better option would be to try and connect in a loop every, say, 5 secs.

File: oak_containers_hello_world_untrusted_app/tests/integration_test.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+fn build_dependencies() -> anyhow::Result<()> {
+    duct::cmd!("just", "all_oak_containers_binaries",)
+        .dir(env!("WORKSPACE_ROOT"))
+        .run()?;
+    Ok(())
+}
+
+fn get_laucnher_args() -> Args {
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/tests/integration_test.rs:65]:
```suggestion
fn get_launcher_args() -> Args {
```

File: oak_containers_hello_world_untrusted_app/tests/integration_test.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        "--stage0-binary={}stage0_bin/target/x86_64-unknown-none/release/stage0_bin",
+        env!("WORKSPACE_ROOT")
+    );
+    let kernel_arg = format!(
+        "--kernel={}oak_containers_kernel/target/bzImage",
+        env!("WORKSPACE_ROOT")
+    );
+    let initrd_arg = format!("--initrd={}/target/stage1.cpio", env!("WORKSPACE_ROOT"));
+    let arg_strs = [
+        "",
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/tests/integration_test.rs:89]: Is this needed?

File: oak_containers_hello_world_untrusted_app/tests/integration_test.rs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        "--port=8080",
+        &system_image_arg,
+        &container_bundle_arg,
+        &vmm_path_arg,
+        &stage0_binary_arg,
+        &kernel_arg,
+        &initrd_arg,
+        "--memory-size=8G",
+        "--ramdrive-size=3000000",
```
Comment (tiziano88) [inline oak_containers_hello_world_untrusted_app/tests/integration_test.rs:98]: Why not create the Args object directly? Unless you specifically want to exercise the flag parsing logic (but that seems unnecessary).

File: oak_containers_launcher/src/qemu.rs [hunk 7]
```diff
@@ -129,49 +122,38 @@ impl Qemu {
...
-                    "ip=10.0.2.10:::255.255.255.0::eth0:off"
-                ]
-                .join(" ")
-            )
+            "-append",
+            [
+                "console=ttyS0",
+                "panic=-1",
+                "brd.rd_nr=1",
+                format!("brd.rd_size={}", params.ramdrive_size).as_str(),
```
Comment (tiziano88) [inline oak_containers_launcher/src/qemu.rs:148]:
```suggestion
                format!("brd.rd_size={params.ramdrive_size}").as_str(),
```

(not sure it works)

File: oak_containers_launcher_server/src/lib.rs [hunk 8]
```diff
@@ -173,7 +177,7 @@ pub async fn new(
     };
     Server::builder()
         .add_service(LauncherServer::new(server_impl))
-        .serve(addr)
+        .serve_with_shutdown(addr, shutdown.map(|_| ()))
```
Comment (tiziano88) [inline oak_containers_launcher_server/src/lib.rs:180]: What is this mapping from? Isn't it already `()`?

**Review by tiziano88 — APPROVED** [review]
very cool

File: oak_containers_system_image/10-enp0s1.network [hunk 9]
```diff
@@ -2,4 +2,4 @@
 Name=enp0s1
 
 [Network]
-Address=10.0.2.10/24
+Address=10.0.2.15/24
```
Comment (andrisaar) [inline oak_containers_system_image/10-enp0s1.network:5]: Why this change?

**Review by andrisaar — APPROVED** [review]

**Review by conradgrobler — COMMENTED** [review]
Thanks

**Review by jul-sh — COMMENTED** [review]
Thank you!! :))
